geomesa-filter/src/main/scala/org/locationtech/geomesa/filter/FilterHelper.scala

```diff
@@ -516,15 +517,31 @@ object FilterHelper {
     */
   def flatten(filter: Filter): Filter = {
     filter match {
-      case and: And =>
-        val (ands, others) = and.getChildren.map(flatten).partition(_.isInstanceOf[And])
-        ff.and(ands.flatMap(_.asInstanceOf[And].getChildren) ++ others)
+      case and: And  => ff.and(flattenAnd(and.getChildren, Seq.empty[Filter]))
+      case or: Or    => ff.or(flattenOr(or.getChildren, Seq.empty[Filter]))
+      case f: Filter => f
+    }
+  }
 
-      case or: Or =>
-        val (ors, others) = or.getChildren.map(flatten).partition(_.isInstanceOf[Or])
-        ff.or(ors.flatMap(_.asInstanceOf[Or].getChildren) ++ others)
+  @tailrec
+  private def flattenAnd(remaining: Seq[Filter], result: Seq[Filter]): Seq[Filter] = {
+    if (remaining.isEmpty) { result } else {
+      val (_remaining, _result) = remaining.head match {
+        case f: And => (remaining.tail ++ f.getChildren, result)
+        case f      => (remaining.tail, result :+ flatten(f))
```

Comment:
not sure if it would be more efficient to use a mutable list and append things to it...